### ISSUE TITLE
How to deal with the legality of the action space

### ISSUE THREAD

**Issue by noc-turne** [aai-institute/tianshou#645]
For example, when playing Go, the output of the neural network can range from 1 to 361,
but with the progress of Go, many points have been occupied, unable to make a move at that point,
so how to deal with this problem?

**Comment by Trinkle23897** [comment]
Using mask from the observation. For example,
```
x..
..x
.x.
```
can be treated as a mask of `[[1, 0, 0], [0, 0, 1], [0, 1, 0]]`. For DQN family, you can simply pick up the action from `np.where(mask == 0)`; for other methods (sampling action from a Categorical distribution), you can apply the mask on logits and make the sample probability of the unavailable action to 0.

**Comment by noc-turne** [comment]
OK. But how can I realize that with tianshou?
Do I need to modify the Collector and how?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "thu-ml/tianshou"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年5月18日(星期三) 晚上9:25
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [thu-ml/tianshou] How to deal with the legality of the action space (Issue #645)





 
Using mask from the observation. For example,
 x.. ..x .x.  
can be treated as a mask of [[1, 0, 0], [0, 0, 1], [0, 1, 0]]. For DQN family, you can simply pick up the action from np.where(mask == 0); for other methods (sampling action from a Categorical distribution), you can apply the mask on logits and make the sample probability of the unavailable action to 0.

**Comment by Trinkle23897** [comment]
This feature has already been supported in DQNPolicy. See https://github.com/thu-ml/tianshou/blob/277138ca5b050518aacaaea367192f910fbe666d/tianshou/policy/modelfree/dqn.py#L130-L144

In short, add an env wrapper to change obs space into a dict space, e.g., original obs is 3x3 np array, and the new obs is a dict like
```python
{
  "obs": obs,
  "mask": mask,
}
```

For policy-gradient, see #542

**Comment by noc-turne** [comment]
Did you mean env.step() returns and env.reset() returns batch_obs?
![image](https://user-images.githubusercontent.com/86636997/169647582-5a3473ea-f968-433c-a8f5-2ce5ff714fbc.png)
But I tried, it seems that collector can't work well.
![image](https://user-images.githubusercontent.com/86636997/169647712-fa052771-5c2b-44ef-9528-870170841d39.png)
I don't know where the problem is.

**Comment by Trinkle23897** [comment]
No. env is a single timestep abstraction. You only need to return the original data (without batch_size). For example,
```python
class EnvWrapper(gym.ObservationWrapper):
  def __init__(self, env):
    super().__init__(env)
    self.observation_space = gym.spaces.Dict(...)
  def observation(self, obs):
    mask = calc_mask_by_obs(obs)
    return {"obs": obs, "mask": mask}

env = EnvWrapper(gym.make("..."))
train_envs = SubprocVectorEnv([lambda: EnvWrapper(gym.make("...")) for _ in range(100)])
```

**Comment by noc-turne** [comment]
Is it necessary to inherit gym.ObservationWrapper ?
because I created my own environment and didn't inherit gym.

**Comment by Trinkle23897** [comment]
That's fine. Try this:
```python
env = EnvWrapper(YourEnv(args))
train_envs = SubprocVectorEnv([lambda: EnvWrapper(YourEnv(args)) for _ in range(100)])
```

BTW, inherit gym is only a few words. It's not hard to do so.
